envisage/entry_point_plugin_manager.py

```diff
@@ -0,0 +1,42 @@
+# (C) Copyright 2007-2023 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+
+try:
+    from importlib.metadata import entry_points, EntryPoints
+except ImportError:
+    from importlib_metadata import entry_points, EntryPoints
+
+from traits.api import Instance, Str
+
+from envisage.plugin_manager import PluginManager
+
+
+class EntryPointPluginManager(PluginManager):
+    """
+    A plugin manager that loads the initial set of plugins from
+    a collection of Python entry points (see importlib.metadata).
+    """
+
+    #: Name of the entry point group that we'll load plugins from, if
+    #: the 'entry_points' trait is not provided.
+    group = Str("envisage.plugins")
+
+    #: Entry points that the plugins will be loaded from. Overrides
+    #: the 'group' trait.
+    entry_points = Instance(EntryPoints)
+
+    def _entry_points_default(self):
+        return entry_points(group=self.group)
+
+    # Protected 'PluginManager' protocol ######################################
+
+    def __plugins_default(self):
+        return [entry_point.load()() for entry_point in self.entry_points]
```

Comment:
Are there additional arguments that need to be passed in here?  I notice that the egg basket plugin manager is injecting the `application` when it creates the plugin instance, for example

Reply:
I've added a test, but it's already passing. I suspect we _do_ need to pass the application here, but I might need to work a bit harder to find a test that exercises the issue.

Reply:
It's possible the application gets injected later - I didn't look that deeply, was just comparing with other code that used extension points to load plugins.

Reply:
So yes, in normal use the application gets set on the plugin manager after plugin manager creation, and that then triggers a handler that goes and injects the application into each of the plugins.

But horrible Traits-initialization-order things can happen if you do something like:
```
            def _plugin_manager_default(self):
                return EntryPointPluginManager(
                    application=self,
                    entry_points=entry_points,
                )
```

In that case, the `application` trait gets set _first_; that triggers computation of the `_plugins` trait on the plugin manager _before_ we set `entry_points`, and we end up picking up from the default group instead of from the given entry points. So some more traits change handlers are needed to make this work properly.